main.tf

```diff
@@ -14,28 +14,38 @@
  * limitations under the License.
  */
 
+locals {
+  tables = { for table in var.tables : table["table_id"] => table }
+}
+
 resource "google_bigquery_dataset" "main" {
   dataset_id    = var.dataset_id
   friendly_name = var.dataset_name
   description   = var.description
   location      = var.location
 
-  default_table_expiration_ms = var.expiration
+  default_table_expiration_ms = var.default_table_expiration
   project                     = var.project_id
   labels                      = var.dataset_labels
 }
 
 resource "google_bigquery_table" "main" {
-  count         = length(var.tables)
-  dataset_id    = google_bigquery_dataset.main.dataset_id
-  friendly_name = var.tables[count.index]["table_id"]
-  table_id      = var.tables[count.index]["table_id"]
-  labels        = var.tables[count.index]["labels"]
-  schema        = file(var.tables[count.index]["schema"])
-  project       = var.project_id
-
-  time_partitioning {
-    type = var.time_partitioning
+  for_each        = local.tables
+  dataset_id      = google_bigquery_dataset.main.dataset_id
+  friendly_name   = each.key
+  table_id        = each.key
+  labels          = each.value["labels"]
+  schema          = file(each.value["schema"])
```

Comment:
Should we take this opportunity to remove the call to `file` and make `schema` more flexible? Users could choose to define schema inline or define schema separately and invoke `file`.

Reply:
@aaron-lane I don't think there is a straight forward way to do so, there is a big issue:
it is not currently possible to use functions in the tfvars (and `schema` is only a field of the `tables` variable) and I assume it's a usual way for such a complex variable `tables` to be specified in some tfvars file. So giving the user ability to have `schema` inline will break existing way with schema.json. Maybe I'm missing some beautiful way to handle that

Reply:
Fair point; I missed that it was part of an object.